src/types/mod.rs

```diff
@@ -1,73 +1,62 @@
 use itertools::Itertools;
+use std::cmp::Ordering;
 use std::collections::{HashMap, HashSet};
 use std::hash::{Hash, Hasher};
 use std::iter;
 
 use super::parser;
 
-/// Represents a human-assigned name.
-pub type Name_ = Option<String>;
-
 /// Represents a unique identity, a [DeBruijn Index].
 ///
 /// [DeBruijn Index]: https://en.wikipedia.org/wiki/De_Bruijn_index
 pub type DeBruijn = usize;
 
-/// Represents the number of arguments an [`Operator`] takes.
-///
-/// [`Operator`]: trait.Operator.html
-pub type Arity_ = usize;
-
 /// Represents a place in a [`Term`]
 ///
 /// [`Term`]: struct.Term.html
 pub type Place = Vec<usize>;
 
-/// A trait for things that act like symbols
-pub trait Symbol: Eq + Hash + Clone {
-    /// Each symbol must have a unique ID associated with it.
-    type ID;
-    /// Return a new `ID` distinct from any existing `ID`.
-    fn new_id(&[&Self]) -> Self::ID;
-    /// Return a human-readable representation of `self`.
-    fn show(&self) -> String;
-    /// Return the `ID` of `self`.
-    fn id(&self) -> Self::ID;
-}
```

Comment:
I'm not sure that removing `Symbol` is the best way forward. It feels important to me that we be able to see clearly that `Variable` and `Operator` values are similar kinds of things. I was thinking that we could use the `Symbol` trait to perform the same function we were planning for an `Atom` enum. Things which implement `Symbol`, like `Operator` and `Variable` would then be able to be gathered as `Symbol`s.

If we do that, then the `Arity` trait might also make sense, as an `Operator` becomes `Symbol` + `Arity`.

I'm not totally sure this is the right path, so could you help me understand why you made the changes you did?

Reply:
We talked about this offline, and you have a good point in saying that there are only two kinds of thing which should be identified as `Atom`, so an enum might make sense. It still feels conceptually important (at least to me), that an `Operator` and a `Variable` are the same kind of thing except that the `Operator` has an arity. That similarity comes from the bottom up rather than being imposed from the top-down as it would via `Atom`. Is there a good way to allow that to show clearly in the design?

Reply:
Thanks for bringing this up, because this is definitely the most meaningful change and certainly warrants discussion. Critically, consider that `Variable`s are numbers with a (maybe) name, but `Operator`s have potentially some other functionally that the end user may want (imagine something like an evaluator, for example). Plus there's no clear reason that an `Operator` should even have an associated number.

If `Operator`s didn't have associated indices, which I think should be the case, the only thing in common between them and `Variable`s is an optional name and that they're both `Atom`s. This was pretty much my reasoning. What do you think?

Reply:
Ah! We've been thinking about `Variable`s similarly (i.e. they are a number with a (maybe) name), which is great.

If the name is optional for `Operator`, the arity alone isn't enough to give an `Operator` a unique identity (e.g.  there can be multiple `Operator`s of arity 1). Something else must provide a unique identity, like an associated identifier. That is, both `Variable`s and `Operator`s are something like integers equipped with extra information: (maybe) names, and in the case of `Operator`, an arity.

I suppose that's what I've been trying (obliquely) to get at with the `Symbol` trait. Both `Variable` and `Operator` values are just symbols (e.g. integers), equipped with extra information. What do you think?

Reply:
Unique identity is already provided by the `Eq` trait bound, so there's no need for an intermediate value to get that. Hence why the `impl PartialEq for Op` relies on just `id` and `arity` (and _not_ on `name`).